cpp/src/arrow/acero/aggregate_node.cc

```diff
@@ -665,14 +680,47 @@ class GroupByNode : public ExecNode, public TracedNode {
     }
     base += segment_keys.size();
     for (size_t i = 0; i < aggs.size(); ++i) {
-      output_fields[base + i] =
-          agg_result_fields[i]->WithName(aggregate_options.aggregates[i].name);
+      output_fields[base + i] = agg_result_fields[i]->WithName(aggs[i].name);
+    }
+
+    return AggregateNodeArgs<HashAggregateKernel>{schema(std::move(output_fields)),
+                                                  std::move(key_field_ids),
+                                                  std::move(segment_key_field_ids),
+                                                  std::move(segmenter),
+                                                  std::move(agg_src_fieldsets),
+                                                  std::move(aggs),
+                                                  std::move(agg_kernels),
+                                                  std::move(agg_src_types),
+                                                  /*states=*/{}};
```

Comment:
Why states is passed in with ScalarAggregate but not HashAggregate?

Reply:
Technically, this is because `GroupByNode` does not take states in [its constructor](https://github.com/apache/arrow/blob/d9e672f40e831dd8f6a73a21b3a9bebfc35c5a45/cpp/src/arrow/acero/aggregate_node.cc#L543-L549) whereas `ScalarAggregateNode` [does](https://github.com/apache/arrow/blob/d9e672f40e831dd8f6a73a21b3a9bebfc35c5a45/cpp/src/arrow/acero/aggregate_node.cc#L257-L265). I don't think there is a real design justification for this difference; I just didn't change the original design here.